scipy/optimize/_constraints.py

```diff
@@ -0,0 +1,620 @@
+        self.hess = hess
+        self.feasible_constr = feasible_constr
+
+
+class NonlinearConstraint:
+    """Nonlinear constraint
+
+    Parameters
+    ----------
+    fun : callable
+        The function defining the constraint.
+
+            fun(x) -> array_like, shape (m,)
+
+        where x is a (n,) ndarray and ``m``
+        is the number of constraints.
+    jac : callable
+        Jacobian Matrix:
+
+            jac(x) -> {ndarray, sparse matrix}, shape (m, n)
+
+        where x is a (n,) ndarray.
+    hess : {callable, None}
+        Hessian matrix of `dot(fun, v)`:
+
+            hess(x, v) -> {LinearOperator, sparse matrix, ndarray}, shape (n, n)
+
+        where x is a (n,) ndarray and v is a (m,) ndarray.
+    kind : tuple
+        Specifies the type of contraint. Options for this
+        parameters are:
+
+            - ("interval", lb, ub): for a constraint of the type:
+                lb <= fun(x) <= ub
+            - ("greater", lb): for a constraint of the type:
+                fun(x) >= lb
+            - ("less", ub): for a constraint of the type:
+                fun(x) <= ub
+            - ("equals", c): for a constraint of the type:
+                fun(x) == c
+
+        where ``lb``,  ``ub`` and ``c`` are (m,) ndarrays and
+        ``x`` is a (n,) ndarray.
+    feasible_constr : {list of boolean, boolean}, optional
+        Specify if the constraint must be feasible along the iterations.
+        If ``True``  all the iterates generated by the optimization
+        algorithm need to be feasible in respect to a constraint. If ``False``
+        this is not needed. A list can be passed to specify element-wise
+        each constraints needs to stay feasible along the iterations and
+        each does not. Alternatively, a single boolean can be used to
+        specify the feasibility required of all constraints. By default it
+        is False.
+    """
+    def __init__(self, fun, jac, hess, kind, feasible_constr=False):
+        self.fun = fun
+        self.jac = jac
+        self.hess = hess
+        self.kind = kind
+        _, _, _, _, _, fun_len = parse_constraint(self.kind)
+        if isinstance(feasible_constr, bool):
+            self.feasible_constr = np.full(fun_len,
+                                           feasible_constr,
+                                           dtype=bool)
+        else:
+            self.feasible_constr = np.array(feasible_constr,
+                                            dtype=bool)
+
+    def to_canonical(self, sparse_jacobian=None):
+        # Parse constraints
+        eq, ineq, val_eq, val_ineq, sign, fun_len = parse_constraint(self.kind)
+        # Get dimensions
+        n_eq = len(eq)
+        n_ineq = len(ineq)
+        # Set variables
+        self.fun_x = None
+        self.x = None
+        self.jac_y = None
+        self.y = None
+
+        def constr_eq(x):
+            if n_eq > 0:
+                if not np.array_equal(x, self.x):
```

Comment:
You have a point there. A function ``np.array_not_equal`` could be more efficient. I did't find such function however and the use of ``!=`` was troublesome for the initial case when ``self.x=None``.